### ISSUE TITLE
LogContext.get_Enrichers throws RemotingException

### ISSUE THREAD

**Issue by yallie** [serilog/serilog#835]
Got this error after updating Serilog to v2.1:

```
Object '/4476135e_0ad6_49be_a1ba_310183038f8c/n9keufbaxkx3rbttwc7s84qw_2.rem' has been disconnected or does not exist at the server.

Server stack trace: 
   at System.Runtime.Remoting.Channels.ChannelServices.CheckDisconnectedOrCreateWellKnownObject(IMessage msg)
   at System.Runtime.Remoting.Channels.ChannelServices.SyncDispatchMessage(IMessage msg)

Exception rethrown at [0]: 
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at System.Runtime.Remoting.ObjectHandle.Unwrap()
   at Serilog.Context.LogContext.get_Enrichers()
   at Serilog.Context.LogContext.GetOrCreateEnricherStack()
   at Serilog.Context.LogContext.PushProperty(String name, Object value, Boolean destructureObjects)
```

ObjectHandle wrapper seems to be the source of the trouble.
Looks like the object handle is disconnected and garbage collected,
it has something to do with remoting lifetime service and leases.

Why is ObjectHandle wrapper used in the first place?
Is it to prevent LCC slots from leaking into another AppDomain?
Perhaps it can be replaced with a custom wrapper class.

**Comment by nblumhardt** [comment]
Hi Alexey, thanks for the report. `LogicalCallContext` and remoting has been an absolute nightmare for this feature, we thought we finally had a bulletproof implementation in 2.0 but it seems we're not so lucky :( // cc @colin-young @DmitryNaumov @gertjvr 

`LogContext` uses `ObjectWrapper` so that when an app using it calls into a secondary `AppDomain`, any enrichers pushed onto the context stack (which may be arbitrary, non-serializable, user-provided objects) remain in the source `AppDomain` and aren't serialized across with the call.

Am I correct in guessing that the application you're working on has made a call into another `AppDomain` at this point, either through remoting directly, or via SQL Server Management Objects?

Using a subclass of `ObjectHandle` that returns a custom `ILease` from `InitializeLifetimeService()` should fix this, as you suggest, but figuring out how to keep the lease alive long enough without leaking the object entirely will need some thought. Anyone have any ideas?

.NET 4.6 introduced `AsyncLocal`, which does not suffer from this problem (it uses `ExecutionContext` rather than `LogicalCallContext` for the state). We already support `AsyncLocal<>` with a build flag; I'm going to create a PR now that adds .NET 4.6 as an explicit target so that it's used instead of this horrendous API.

Which .NET Framework version are you running, Alexy?

**Comment by yallie** [comment]
>Am I correct in guessing that the application you're working on has made a call into another AppDomain at this point, either through remoting directly, or via SQL Server Management Objects?

Yes, we use remoting. Are you sure that the problem pops up only in remoting scenarios?

>Using a subclass of ObjectHandle that returns a custom ILease from InitializeLifetimeService() should fix this, as you suggest...

Ah, one more thing: a custom wrapper class or ObjectHandle descendant will need Serilog assembly in the target AppDomain to be deserialized. What if Serilog is not deployed there?

>figuring out how to keep the lease alive long enough without leaking the object entirely will need some thought. Anyone have any ideas?

If you're OK with custom wrapper class, then I'd suggest infinite lease time + explicit disconnection via RemotingServices.Disconnect, something like this:

```
// The original idea is by Nathan Evans, I believe
public class DisposableMarshalByRefObject : MarshalByRefObject, IDisposable
{
	public sealed override object InitializeLifetimeService()
	{
		return null;
	}

	private bool Disposed { get; set; }

	public void Dispose()
	{
		if (!Disposed)
		{
			Disposed = true;
			RemotingServices.Disconnect(this);
		}
	}
}
```
So now it's a questions of where and when the wrapper should be disposed of. Perhaps it's when the stack of enrichers goes empty?

>Which .NET Framework version are you running

I'm running .NET 4.6 so using AsyncLocal is not a problem. But I'm a bit confused about ExecutionContext... I believe it includes logical call context as a part of it, doesn't it?

**Comment by nblumhardt** [comment]
Thanks for the great reply, appreciate the feedback.

> Yes, we use remoting. Are you sure that the problem pops up only in remoting scenarios?

I believe the problematic `ObjectHandle` lease only starts after a cross-appdomain call. All 2.0+ builds include this code, so the situations in which it becomes an issue must be fairly limited or we'd be flooded with reports by now (see more info further below).

> Ah, one more thing: a custom wrapper class or ObjectHandle descendant will need Serilog assembly in the target AppDomain to be deserialized. What if Serilog is not deployed there?

Only if the target appdomain uses Serilog's `LogContex`, otherwise the `ObjectHandle` is never `Unwrap()`ped.

> If you're OK with custom wrapper class, then I'd suggest infinite lease time + explicit disconnection via RemotingServices.Disconnect, something like this:

The copying behaviour of `LogicalCallContext` will make it hard for one single thread, that might be emptying the `LogContext`, to be sure that the `ObjectHandle` isn't still active on another logical thread. Or at least, I'd be cautious asserting anything about when the `ObjectHandle` can be disposed, because there are so many potential (mis-)usage patterns, and this API is a pain to work with.

> I'm running .NET 4.6 so using AsyncLocal is not a problem. But I'm a bit confused about ExecutionContext... I believe it includes logical call context as a part of it, doesn't it?

`ExecutionContext` is there in 4.5 but the members supporting `AsyncLocal` aren't (AFAICT).

I'll merge the 4.6 fix now and leave this issue open for discussion.

**Comment by nblumhardt** [comment]
@yallie build `2.2.1-dev-00697` including the 4.6 fix is now on NuGet; it's otherwise close to identical to what is on `master`, so if you can confirm it addresses the issue for you we can push it through to a release build posthaste.

**Comment by yallie** [comment]
@nblumhardt thanks a lot! I built Serilog from your fork when you checked in the 4.6 fix and everything seems to work fine indeed.

>I believe the problematic ObjectHandle lease only starts after a cross-appdomain call.

Ah, now I see, you must be right. Thanks again!

**Comment by DmitryNaumov** [comment]
@nblumhardt @yallie guys, I would appreciate if you describe me what happened, because I don't understand really. In my understanding, ObjectHandle can't be disposed in primary domain as it is stored in LCC. The exception looks like someone from _secondary_ domain attempted to use LogContext. I do remember  that in one of the versions I tried to give a random slot name, so no one from other domains than primary will be able to unwrap ObjectHandle because it won't work anyway, unfortunately I've decided that it is not necessary. I'm talking about this line:

`static readonly string DataSlotName = typeof(LogContext).FullName;`

To summarize my understanding: the attempt to use LogContext was made from secondary AppDomain. Please correct and explain me if I'm wrong.

**Comment by yallie** [comment]
>The exception looks like someone from secondary domain attempted to use LogContext.

I think you're right, the call goes through proxies. I believe random slot name should be ok: an ObjectHandle will leak into other domains, but no one will try to access it.

**Comment by DmitryNaumov** [comment]
I've found it, it was not that random, but good enough to prevent the problem. Even the text was extremely precise, it my mistake it was reverted afterwards.

 Use per AppDomain slot name, so other domain won't accidentally unwrap object reference.
https://github.com/serilog/serilog/pull/733/commits/3b55d2dbcefefc97fe0db37258195190bb4a9d5b

**Comment by nblumhardt** [comment]
@DmitryNaumov that looks like the way to go, thanks for the insight. We could just do something like:

```csharp
static readonly string DataSlotName = typeof(LogContext).FullName + "@" Guid.NewGuid();
```

?

**Comment by DmitryNaumov** [comment]
@nblumhardt if you don't need display friendly name (I think you don't), then that's fine. I still hope we did bullet-proof remoting version and this is just a small mistake.

**Comment by nblumhardt** [comment]
Righto, had a shot at fixing it in #840 - though the `AppDomain.CurrentDomain.Id` was being appended, it's not sufficient to create a unique identifier when inter-process remoting takes place.
